### PR TITLE
test(VIOL-0016): cover current_item GUID guard and source_mapping fallback

### PR DESCRIPTION
**Pull request by Muizzkolapo**
## Summary
- fix(VIOL-0016): preserve previous-stage lineage in LineageEnricher
- fix(VIOL-0016): broaden lineage fix coverage, O(1) lookup, batch + DataGenerator symmetry
- fix(VIOL-0016): address review followups
- fix(VIOL-0016): add direct _get_parent_item unit tests
- fix(VIOL-0016): guard parent_records against lineage-less records
- fix(VIOL-0016): guard current_item GUID match and add source_mapping fallback
- test(VIOL-0016): cover batch seed-record path
- test(VIOL-0016): cover current_item GUID guard and source_mapping fallback

## Verification
- ruff + pytest pass (.harness/ci.ok)
- pi-consensus pass (.harness/review.ok)
- smoke pass (.harness/smoke.ok)

### CODE REVIEW

File: agent_actions/processing/enrichment.py [hunk 1]
```diff
@@ -134,18 +150,51 @@ def enrich(self, result: ProcessingResult, context: ProcessingContext) -> Proces
         result.node_id = base_node_id
         return result
 
+    @staticmethod
+    def _with_parent_fallback(item: dict, parent_index: dict[str, dict] | None) -> dict:
+        """If *item* lacks lineage, look up a richer match in parent_index by source_guid."""
+        if item.get("lineage") or item.get("node_id"):
```
Comment (Muizzkolapo) [inline agent_actions/processing/enrichment.py]: 🟡 **Correctness (latent) — inclusion predicate is broader than the lineage consumer can use.** parent_records is filtered by `r.get("lineage") or r.get("node_id")` (here, and in the 3 call sites), so a record with `node_id` but **no `lineage` key** qualifies as a parent. But `add_unified_lineage`/`build_lineage` gate on `"lineage" in parent_item` → a node_id-only parent yields `lineage=[child_node_id]` (self-only), silently dropping the chain — the exact thing this PR fixes. Worse, because parent_index is consulted before source_data, such a record can *shadow* a lineage-bearing source_data record for the same guid. I couldn't find a production path that feeds a node_id-without-lineage record into a non-first-stage parent today (so it's latent), but the predicate should match what's consumed (filter on `lineage`, or have the builder fall back to the parent's `node_id`). The new tests only use parents carrying **both** fields, so this gap is uncovered.

File: agent_actions/processing/enrichment.py [hunk 2]
```diff
@@ -80,13 +90,15 @@ def enrich(self, result: ProcessingResult, context: ProcessingContext) -> Proces
...
             ):
-                # Index-based lookup — resolve parent by source_mapping
                 source_idx = result.source_mapping[i]
                 if isinstance(source_idx, list):
                     # Many-to-one: multiple input records merged into one output
                     source_items = [
                         context.source_data[idx] for idx in source_idx if idx < source_data_len
                     ]
+                    source_items = [
+                        self._with_parent_fallback(s, parent_index) for s in source_items
```
Comment (Muizzkolapo) [inline agent_actions/processing/enrichment.py:87]: 🟡 **Correctness (low) — many-to-one fallback can duplicate `lineage_sources`.** When two merged `source_idx` entries share a `source_guid`, `_with_parent_fallback` resolves both to the **same** parent_index object; `add_lineage_tracking_from_sources` then accumulates that parent's chain once per occurrence → e.g. `lineage_sources=['root','p','root','p']`. Pre-PR the raw seeds had no lineage so this never duplicated. Dedup the parents (or the resulting `lineage_sources`).

**Review by Muizzkolapo — COMMENTED** [review]
## Code review — PR #744 (VIOL-0016): wire `parent_records` through the lineage enricher

Max-effort review: 10 finder angles → adversarial 1-vote verification → gap sweep. 12 findings posted inline.

**Overall:** the core wiring is sound. The two strongest-looking *regression* hypotheses were **refuted** under tracing — the batch `current_item` source_guid guard is a no-op (the reconciler's `get_source_guid` reads the same `original_row['source_guid']`), and the unwired `processing.py` batch-results context is only used for error results that were already parentless on main. No **live** correctness regression survived. What remains:

**Correctness (latent / edge):**
1. 🟡 The parent inclusion predicate (`lineage OR node_id`) is broader than what `add_unified_lineage` consumes (requires `lineage`), so a `node_id`-only parent silently produces self-only lineage — and the new tests never cover that shape. (enrichment.py:156)
2. 🟡 Many-to-one fallback can duplicate `lineage_sources` when merged sources share a guid. (enrichment.py:100)
3. 🟡 Per-item lookup reads the expansion-regenerated `source_guid`, missing the parent. (enrichment.py:139)
4. 🟡 `parent_index` precedence can shadow a more-complete `source_data` chain (mostly by-design). (enrichment.py:190)

**Scope / hygiene (most actionable before merge):**
- 🟠 `uv.lock` is off-scope: it **downgrades** `revision 3→2` vs main and bundles a version bump + dependency-bound bumps under a `test(...)` PR. Revert or split it out. (uv.lock:2)

**Cleanup / altitude:** the lineage-bearing predicate is duplicated 4× across 3 construction sites (leaky — a missed site = silent lineage regression); a dead-in-practice linear-scan branch and a redundant batch default; `_get_parent_item` juggles four candidate-parent representations; and `_with_parent_fallback` is a second resolver that ignores `current_item` + the new guard.

**Verified clear (not flagged):** stale-`parent_item` reuse (source_guid is always `None` when source_mapping is set → per-item recompute always runs); last-wins/first-wins divergence (the two branches are mutually exclusive); `ProcessingContext` field-order and `dataclasses.replace` propagation.

_Automated review via `/code-review` (max effort)._

File: agent_actions/processing/enrichment.py [hunk 3]
```diff
@@ -123,7 +137,9 @@ def enrich(self, result: ProcessingResult, context: ProcessingContext) -> Proces
                         parent_item = None
             elif use_per_item_parent_lookup:
                 item_source_guid = item.get("source_guid")
```
Comment (Muizzkolapo) [inline agent_actions/processing/enrichment.py:126]: 🟡 **Correctness (latent, pre-existing) — per-item lookup reads a regenerated `source_guid`.** For expansions the loop overwrites `item["source_guid"]` with a fresh UUID at lines 81-84 (saving the old one to `parent_source_guid`). This per-item branch then reads `item.get("source_guid")` — the *new* UUID — so `parent_index`/`source_index` (keyed on the upstream guid) miss → self-only lineage. Reachable only with `is_expansion` + `result.source_guid=None` + no source_mapping (not produced today), but the new `parent_index` lookup inherits the wrong key. Consider looking up by `parent_source_guid` when present.

File: agent_actions/processing/enrichment.py [hunk 4]
```diff
@@ -134,18 +150,51 @@ def enrich(self, result: ProcessingResult, context: ProcessingContext) -> Proces
...
         if context.is_first_stage or not source_guid:
             return None
 
         if context.current_item:
-            return context.current_item
+            current_guid = context.current_item.get("source_guid")
+            if current_guid is None or current_guid == source_guid:
+                return context.current_item
+
+        if parent_index is not None:
```
Comment (Muizzkolapo) [inline agent_actions/processing/enrichment.py:190]: 🟡 **Correctness (low, mostly by-design) — parent_index shadows a possibly-more-complete source_data parent.** `parent_index` is checked before `source_data`. In `pipeline._process_by_strategy`, `parent_records` comes from `data` while `source_data` may be independently reloaded from storage; for a shared guid, a parent_records record with a *shorter* lineage now wins over a longer source_data chain that pre-PR would have been used. This is the PR's intended precedence (asserted by `test_parent_records_preferred_over_source_data`), so flag only as an edge to confirm: when both carry the same guid, is parent_records always the more-complete chain?

File: agent_actions/processing/enrichment.py [hunk 5]
```diff
@@ -134,18 +150,51 @@ def enrich(self, result: ProcessingResult, context: ProcessingContext) -> Proces
         result.node_id = base_node_id
         return result
 
+    @staticmethod
+    def _with_parent_fallback(item: dict, parent_index: dict[str, dict] | None) -> dict:
```
Comment (Muizzkolapo) [inline agent_actions/processing/enrichment.py:150]: 🟢 **Altitude — two parent-resolution paths can disagree.** `_with_parent_fallback` (the source_mapping path) resolves a parent purely by `source_guid` in `parent_index`; it ignores `current_item` and the new source_guid guard that `_get_parent_item` applies. So a record routed through the source_mapping branch vs the `_get_parent_item` branch can resolve to a *different* parent for identical inputs — the kind of split that makes lineage bugs non-reproducible. Consider one shared resolver used by both paths.

File: agent_actions/prompt/data_generator.py [hunk 6]
```diff
@@ -67,12 +67,19 @@ def create_agent_with_data(
             GenerationError: If agent creation or data generation fails.
         """
         try:
+            source_content_list = source_content if isinstance(source_content, list) else []
+            parent_records = [
+                r
+                for r in source_content_list
+                if isinstance(r, dict) and (r.get("lineage") or r.get("node_id"))
```
Comment (Muizzkolapo) [inline agent_actions/prompt/data_generator.py]: 🟢 **Reuse / Altitude — the "record carries lineage" predicate is duplicated 4× and derived at 3 call sites.** `isinstance(r, dict) and (r.get("lineage") or r.get("node_id"))` appears here, in `pipeline.py:481`, `batch_context_adapter.py:27`, and (as truthiness) `enrichment.py:156`, with no shared helper in `utils/lineage/`. Computing `parent_records` at three separate construction sites is leaky: a new subsequent-stage `ProcessingContext` that forgets to wire it silently regresses to source_data-only lineage with no error. Extract `LineageBuilder.is_lineage_bearing(record)` and, ideally, derive parent_records at one seam.

File: tests/unit/core/test_lineage_parent_records.py [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+            source_data=[seed],
+            parent_records=[lineage_parent],
+        )
+        enriched = LineageEnricher().enrich(result, context)
+        enriched_item = enriched.data[0]
+
+        assert enriched_item["lineage"][0] == "stage_one"
+        assert enriched_item["lineage"][-1] == enriched_item["node_id"]
+
+    def test_many_to_one_source_mapping_recovers_lineage_for_each_source(self):
```
Comment (Muizzkolapo) [inline tests/unit/core/test_lineage_parent_records.py:348]: 🟡 **Test quality — weak coverage for the merge path and the predicate.** This is the only many-to-one fallback test, and it asserts only `"p_a" in lineage` (membership) with two *distinct* parents — so it can't catch duplicate `lineage_sources` (see enrichment.py:100) or a wrong base-lineage order. Separately, no test uses a parent that has `node_id` but no `lineage` (the gap at enrichment.py:156): every parent fixture here carries both fields, so the predicate-vs-consumer mismatch is untested. Add a shared-guid fan-in case and a node_id-only parent case.

File: agent_actions/processing/enrichment.py [hunk 8]
```diff
@@ -134,18 +150,51 @@ def enrich(self, result: ProcessingResult, context: ProcessingContext) -> Proces
...
-            return context.current_item
+            current_guid = context.current_item.get("source_guid")
+            if current_guid is None or current_guid == source_guid:
+                return context.current_item
+
+        if parent_index is not None:
+            hit = parent_index.get(source_guid)
+            if hit is not None:
+                return hit
+        elif context.parent_records:
```
Comment (Muizzkolapo) [inline agent_actions/processing/enrichment.py]: 🟢 **Simplification — dead-in-practice branch.** `enrich()` always builds `parent_index` whenever `parent_records` is non-empty (lines 52-58) and always passes it to `_get_parent_item` (lines 62-64, 140-142). So `parent_index is not None` whenever `parent_records` is truthy, making this `elif context.parent_records:` linear scan unreachable from production — only the direct unit tests hit it. Have the method accept/build one representation and drop the elif (its first-match semantics also differ from the dict's last-wins, a latent trap if it ever does run).

File: agent_actions/processing/batch_context_adapter.py [hunk 9]
```diff
@@ -19,14 +19,23 @@ def to_processing_context(
         original_row: dict[str, Any],
         record_index: int,
         output_directory: str | None = None,
+        parent_records: list[dict[str, Any]] | None = None,
     ) -> ProcessingContext:
         """Build a ProcessingContext from batch-side state."""
+        if parent_records is None:
```
Comment (Muizzkolapo) [inline agent_actions/processing/batch_context_adapter.py:29]: 🟢 **Simplification — redundant default in the batch path.** `parent_records=[original_row]` is never consumed here: `_get_parent_item` returns `current_item` (=`original_row`, same guid) first, and the `parent_index`/`_with_parent_fallback` consumers only fire when `source_data`+`source_mapping` are set — which neither batch caller (`batch_result_strategy.py:376`, `:544`) passes. Drop the 7-line block and default `parent_records=[]`; a reader otherwise has to trace three lookup branches to learn it has no effect.

File: agent_actions/processing/enrichment.py [hunk 10]
```diff
@@ -134,18 +150,51 @@ def enrich(self, result: ProcessingResult, context: ProcessingContext) -> Proces
...
+            return item
+        if parent_index is None:
+            return item
+        sg = item.get("source_guid")
+        if sg is None:
+            return item
+        richer = parent_index.get(sg)
+        return richer if richer is not None else item
+
     def _get_parent_item(
```
Comment (Muizzkolapo) [inline agent_actions/processing/enrichment.py:164]: 🟢 **Altitude — `_get_parent_item` juggles four candidate-parent representations.** `current_item`, `parent_index`, `context.parent_records`, and `source_index`/`context.source_data` are four inputs behind a hand-ordered precedence ladder, where index-vs-list pairs are two encodings of the same set kept in sync only by `enrich()`. Since `enrich()` is the sole production caller, consider resolving the ordered candidate set once upstream and passing the enricher a single representation — the dual params exist mainly to let tests bypass index construction.

File: agent_actions/processing/enrichment.py [hunk 11]
```diff
@@ -49,9 +49,19 @@ def enrich(self, result: ProcessingResult, context: ProcessingContext) -> Proces
                 if (sg := item.get("source_guid")) is not None
             }
 
+        parent_index: dict[str, dict] | None = None
```
Comment (Muizzkolapo) [inline agent_actions/processing/enrichment.py]: 🟢 **Reuse — index comprehension duplicated.** `{sg: item for item in X if (sg := item.get("source_guid")) is not None}` is now written twice in `enrich()` (source_index at 45-50, parent_index at 52-58), identical shape. Extract `_index_by_source_guid(records)` so the two indexes can't drift (e.g. one later changing duplicate-guid handling).

File: uv.lock [hunk 12]
```diff
@@ -1,5 +1,5 @@
 version = 1
-revision = 3
+revision = 2
```
Comment (Muizzkolapo) [inline uv.lock]: 🟠 **Conventions / scope — `uv.lock` is off-scope and downgrades the lock revision.** This PR is titled `test(VIOL-0016): ...` but `uv.lock` reverts `revision = 3` → `revision = 2` (origin/main is on revision 3) and bundles unrelated changes: project version `0.1.16`→`0.2.6`, and dep-bound bumps (cohere `<7`→`<8`, google-genai `<2`→`<3`, mypy `<2`→`<3`, pytest dedup). This breaks repo `CLAUDE.md` *Git and Change Hygiene* ("Keep commits atomic … avoid 'misc fixes' bundles"; "Treat generated files carefully") and `CLAUDE.local.md` ("One branch, one PR. Don't pile multiple unrelated changes into one ship."). The revision downgrade also risks a lockfile-format regression vs main. Revert the `uv.lock` changes from this PR (or split them out) and regenerate against current main.

**Comment by Muizzkolapo** [comment]
Addressed in be33e21e:

**Acted on:**
- 🟠 `uv.lock` reverted to origin/main. An earlier commit on this branch accidentally bundled a project version bump (0.1.16 → 0.2.6), several `requires-dist` upper-bound widenings, and a lockfile revision downgrade (3 → 2) that have nothing to do with this fix.
- 🟡 `enrichment.py:156` predicate tightened. The parent-record filter required `lineage OR node_id`, but `LineageBuilder.add_unified_lineage` only extends the chain when `"lineage" in parent_item`. A node_id-only record could qualify as parent and fall through to self-only lineage — re-introducing the bug. Filter is now `lineage` only at all four construction sites (`pipeline.py`, `data_generator.py`, `batch_context_adapter.py`, `_with_parent_fallback`). New test pins the node_id-only case.

**Deferred (not regressions, scope-creep for this PR):**
- 🟡 enrichment.py:100 lineage_sources duplication on shared-guid fan-in. Edge case; would need a builder-level dedup.
- 🟡 enrichment.py:139 expansion source_guid regeneration shadows parent lookup. Pre-existing, not introduced by this branch.
- 🟢 Altitude cleanups: extract `is_lineage_bearing` helper, deduplicate the `_index_by_source_guid` comprehension, drop the dead linear-scan branch, unify `_with_parent_fallback` and `_get_parent_item`, drop the redundant batch default. All real but each one is its own small PR; bundling them would make this diff harder to review for the actual bug fix.

Filing the deferred items as follow-ups rather than expanding scope.

**Comment by Muizzkolapo** [comment]
Round 2 of follow-ups in 9707cd6f. All 🟢 altitude/cleanup items now addressed:

- **Predicate duplicated 4×** → `LineageBuilder.is_lineage_bearing` used at every parent_records construction site (pipeline, DataGenerator, BatchContextAdapter, `_with_parent_fallback`).
- **Index comprehension duplicated** → `_index_by_source_guid` static method; `source_index` and `parent_index` now share one builder.
- **Dead `elif` linear scan** + **trailing linear scan over `source_data`** → both removed. `_get_parent_item` lazily builds the missing index, so it has one resolution path per source (parent_records, source_data) with consistent dict semantics.
- **Redundant batch default `parent_records=[original_row]`** → dropped. `current_item` is the explicit batch parent; the default was never consumed by `_get_parent_item` or `_with_parent_fallback`. Batch tests updated to reflect the new contract.

7772 tests pass, mypy clean.

Still deferred (genuine cross-PR work):
- enrichment.py:100 `lineage_sources` dedup on shared-guid fan-in — touches `LineageBuilder.add_lineage_tracking_from_sources`
- enrichment.py:139 expansion `source_guid` regeneration — pre-existing, not introduced by this branch
- `_with_parent_fallback` vs `_get_parent_item` divergence — would unify the two resolvers into one; the cleanups here narrowed the gap but didn't merge them
